docs/foreign-keys.md

```diff
@@ -0,0 +1,199 @@
+            "type": "string",
+            "format": "select",
+            "enumSource": [["Will be dynamically overwritten"]]
+        },
+        "fieldType": {
+            "options": {
+                "hidden": true
+            },
+            "type": "string",
+            "enum": ["reference"]
+        }
+    }
+}
+```
+
+Then when loading a form for the schema editor, the `schemas-service` script
+could populate the `referenceTarget` with all available RecordType UUIDs.
+Additional care would have to be taken to make sure that the data collector
+view would only show options for Records that belong to this RecordType, but
+that filtering should be trivial. 
+
+#### Requirements
+
+For Record-to-Record foreign keys to be workable, they need to permit three basic
+operations:
+
+1. **Lookups**: Foreign key relationships need to support nested search and
+   filtering to match the rest of the Grout field types.
+2. **Validation**: When an incoming record includes a foreign key, the validation function
+   (`grout.models.SchemaModel.validate_schema`) needs to be able to check
+   whether the referent exists in the database.
+3. **Indexing**: Database lookups based on foreign keys must be performant.
+   Without a way to index foreign key relationships, lookups may not be
+   reasonable in large production systems. 
+
+#### A brief note on null pointers
+
+One operation that didn't make this list is **Resolution of null pointers**.
+Null pointers are a classic problem with non-relational data stores: if document
+A points to document B, and then document B is removed from the system, NoSQL
+databases offer no system for automatically resolving the null reference
+(whether by raising an error, setting a default, or simply storing a `NULL`
+value for the foreign key). In the example given above, we could imagine a user
+deleting an Event referenced by a Poster, and then the UUID stored in
+Poster.event would not reference a coherent entity in the database.
+
+Null pointers are indeed a problem with Record-to-Record foreign keys, but this
+problem is already endemic to Grout: if a field is removed from a schema during
+an upgrade to a new version (i.e. a definition of a new RecordSchema), Grout
+silently permits the field to remain but become unaccessible. From the reverse
+perspective, if a field is _added_ in the new schema, old records will not store
+values for that field, and so will not appear in the relevant filters.
+
+While not precisely a "null pointer," the issue is functionally the same: some data
+in the system is silently pointing to nothing. While null reference is a serious
+issue, it is not specific to foreign
+keys in this context, and it will need to be addressed through the user
+interface in a way that is common to all null reference problems in Grout.
+
+### Evaluating the requirements
+
+#### Lookups
+
+Since Record-to-Record foreign keys would store UUIDs (as strings) for the
```

Comment:
I'd like to see some more detail here; we do some queries in DRIVER that are quite a bit more complex than the API lookups that the base Grout API offers. I don't think we necessarily _need_ to be able to accomplish exactly the same queries with external references as we can accomplish with internal references, but it would be helpful to know where the boundaries are.

In particular, the query I'm thinking of is for generating some of the analysis that DRIVER provides: https://github.com/WorldBank-Transport/DRIVER/blob/master/app/data/views.py#L332

That query is a bit dense, but the gist of it is that it allows you to ask questions like "for each different vehicle type, how many crashes involved vehicles of that type during each of the past 12 months?" Currently, that can be done with a single (complicated) query (well, two, I think, one to get the schema up front). You could imagine wanting to ask a similar question about external references as well, for example, "For each Event genre (assuming you have some sort of classification system for Events), how many posters were recorded over each of the past 12 months?"

The reason I'm asking this is that I think a lot of people reach for a normalized data model by instinct, whereas Grout currently only offers a denormalized data model. If we add normalized data model capabilities, I think one thing that'll be important is to have a clear understanding for when to choose one over the other, and knowing if there are any limitations on the analysis you can do with normalized data will be a big part of developing that understanding.

Reply:
I appreciate the thoroughness here -- I breezed past this section sort of assuming it was solved, so it's helpful to have some concrete problems to look at. I don't have much to say about it now since I'm still getting my head around the particular problem, but I agree that knowing the limits of this approach for querying ahead of time is critical.